Improve net cdf functionality

CaitlinA opened the pull request:
Hi @dschlaep,

Can you take a look at the documentation I've done this far?

* check documentation for sensibility
* all checks passed for including a vignette in the repo
* idea for dummy example date for calculate_nominal_resolution function?

I didn't write additional documentation for create_netCDF_from_raster_with_variables or read_netCDF_to_raster functions. Would you like me to? Should I included these function in the vignette? If yes, can you provide an idea for dummy example data (maybe the same as calculate_nominal_resolution). 

I have a couple of to dos to finish up, but I thought it would be good for you to take a look through when you have the time.

Thanks!

CaitlinA committed:
Delete undeveloped create_netCDF_from_raster_with_time function

CaitlinA committed:
Split create_netCDF_from_array_with_variables function into 2 new funcs

CaitlinA committed:
Add options for setting spatial elemnts to create_raster_from_variables

Function can look for spatial information for raster in two ways:
1) From rSFSW2 SFSW2_prj_meta objects
2) locations and CRS

CaitlinA committed:
hash out get_raster_datatype (temporary)

convert the dataType in the raster with the function get_raster_datatype is causing the RasterBrick that is returned to be fauilt. Temporarily hashed out while I diagnose this issue.

CaitlinA committed:
Add spatial arguments to create_empty_netCDF_file

* Users can now enter a grid (aka a raster) OR just points and a CRS
- Uncertain if these arguments will work on data that isn't gridded.

* Rename 'x' arguments to 'data' argument

CaitlinA committed:
Add functionality to create_empty_netCDF_file

This function now creates empty netcdf file for 3D arrays organized by variables or time.

organized by depth is partially scoped.

CaitlinA committed:
Changes to create_empty_netCDF_file

* added back time_bounds argument
* added depth_bounds argument
* add isGridded argument
* fixed time dimension testing
* more checks for proper spatial arguments

CaitlinA committed:
Improved depth logic to create_empty_netcdf

* function can accept variables, time, depth, or time and depth
* if time and depth are both present, depth is assumed to be the third dimension

CaitlinA committed:
Minor improvements

CaitlinA committed:
Functionality to populate_netcdf_from_array

Added major functionality to 'populate_netcdf_from_array'

Minor changes to 'create_empty_netcdf_file'

CaitlinA committed:
Added grid argument back to populate_netcdf_from_array

* Always need locations argument, but now can have a empty grid

* deleted some unnecessary stuff from the code

CaitlinA committed:
Add documentation to create_empty_netCDF_file and populate_netCDF_from_array functions.

CaitlinA committed:
Updates to documentation

.Rd files for functions
Delete references to old functions

CaitlinA committed:
Documentation create_empty_netcdf_file

* changes to documentation
* added LOTS of detail to an example 1

CaitlinA committed:
Documentation expanded

updated documentation and examples for create_netcdf and populate_netcdf functions.

Create draft of netcdf vignette

CaitlinA committed:
Merge branch 'master' into ImproveNetCDFFunctionality

CaitlinA committed:
delete create_netCDF_from_raster_with_variables

CaitlinA committed:
Add information about dimensions to vignette

CaitlinA committed:
Update documentation

CaitlinA committed:
Fix errors and warnings in check

CaitlinA committed:
Fixed functionality in package accounting for multiple variables

Added example to vignette with mult. vars

dschlaep commented:
@CaitlinA 

Running check, e.g., `devtools::check()`, on 90ac802737a17933be857812f144a635d525ddf5
```
...
> checking package dependencies ... NOTE
  Package suggested but not available for checking: ‘rmarkdwon’

...

> checking R code for possible problems ... NOTE
  create_raster_from_variables: no visible global function definition for
    ‘SpatialPoints’
  create_raster_from_variables: no visible global function definition for
    ‘proj4string<-’
  create_raster_from_variables: no visible global function definition for
    ‘CRS’
  create_raster_from_variables: no visible global function definition for
    ‘extent<-’
  create_raster_from_variables: no visible global function definition for
    ‘extent’
  Undefined global functions or variables:
    CRS SpatialPoints extent extent<- proj4string<-

> checking for non-standard things in the check directory ... NOTE
  Found the following files/directories:
    ‘dummynetcdf.nc’

── Test failures ─────────────────────────────────────────── spelling ────

> if (requireNamespace("spelling", quietly = TRUE)) {
+ 
+   spelling::spell_check_test(
+     vignettes = TRUE,
+     error = TRUE,
+     skip_on_cran = TRUE
+   )
+ }
Error: Potential spelling errors: Attibute, CCSM, CRS, CV’s, DRS, IPSL, LorA, NetCDFs, RangeResil, SEUG, SpatialPoints, WDD, XXXXYYZZ, YearMonthDay, allFilesPerVariable, attribtues, calender, conjuction, dataType, degined, endT, gn, gridded, nc, netcdf, ppt, pre, rgdal, startT, strucutured, vals, variableName, variantLabel, veritical, wtih
Execution halted

1 error x | 3 warnings x | 5 notes x
```

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+This vignette is an overview of how to use the functions in the rSW2analysis package to
+convert arrays of rSOILWAT2 and SOILWAT2 output data into netCDFs that adhere to
+[CF conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#_orthogonal_multidimensional_array_representation).
```
Why does this link to `Orthogonal multidimensional array representation`? Why not to `http://cfconventions.org/`?

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+## Variable Data
+
+The netCDF functions in this package expect an array, in which each row represents
+data for a site. Columns represent one of three things:
```
data for a site --> data for a site or gridcell

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+* multiple different variables for one site
+* one variable with multiple measurements (time series)
+* one variable with multiple depth measurements
```
multiple depth --> multiple soil depth
or, maybe even clearer: one variable with a vertical component, e.g., multiple soil depth measurements

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+be either be organized on a regularly spaced grid (i.e. wall to wall simulations for
+a given area) or not (i.e. project data representing plot points in a park).
+
+If the former, the coordinates need to be in order. If the later,
```
Not clear what line 64 means. Does this mean that the coordinate values should be monotonically increasing?
It also appears that the "If the later, " starts and doesn't finish a sentence?

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+All NetCDFs will have at least two dimensions, latitude and longitude.
+
+The bounds and values of the latitude and longitude are determined from the location
```
The bounds of the spatial axes are only inferred from the location data in the case of a gridded` datasets, no?

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+The values refer to the midpoint of each interval of the time dimension.
+
+The bounds and values need to be in the same units.
```
I still think we need to highlight that our function expects a vector instead of a matrix for the bounds, despite the fact that bounds are two-dimensional; and we should explain that the expected organization of that vector is `c(start1, end1, start2, end2, ..., startn, endn)` despite the repetition due to `endk == startk+1` -- and not `c(start1, end1, end2, ..., endn)`.

See http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#cell-boundaries: "A boundary variable will have one more dimension than its associated coordinate or auxiliary coordinate variable. " and "Note that the boundary variable for a set of N contiguous intervals is an array of shape (N,2). "

CaitlinA replied:
fixed

dschlaep replied:
I don't understand what `The bounds of the time dimension are the start and end of each interval of the time dimension.` means without knowing what it should mean.

What about: "The boundary values for a continuous time coordinate are given by the concatenated vector of start/end pairs of each time interval, i.e., c(start1, end1, start2, end2, ..., startn, endn)."

CaitlinA replied:
fixed.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
````diff
@@ -0,0 +1,400 @@
+```{r, depthD}
+
+    ## depth vals
+    depths_max <- c(5, 10, 15) # maximum depth of each layer of depth
````
We require that the variables representing spatial and time coordinates axes have interval mid-points as values; however, this here suggests that the variable representing the vertical coordinate axis uses interval end-points as values. Just make it clear that this is different (for SOILWAT2 reasons).

CaitlinA replied:
Tried to clarify. See lines 84 - 101 and 116 - 153.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+#### Nominal resolution
+
+If the data is calculated and reported on a grid, CF convention calls for the
```
It's not in the CF-1.8 convention; I'm pretty sure that "nominal resolution" is a CMIP6 standard attribute. See "Appendix 2: Algorithms for Defining the “nominal_resolution” Attribute" (https://docs.google.com/document/d/1h0r8RZr_f3-8egBMMh7aqLwy3snpD6_MrDz1q8n5XUk)

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+where:
+
+* var – Shorthand name for variable that is contained in files (i.e. ppt for precipitation, WDD for wet degree days)
+* LorA – Land (L) or atmospheric (A) data
```
Can we change the `LorA_timestep` part be more similar to what CMIP uses? e.g., Lmon, day, Amon, 
And also mention `fx` for time-invariant variables, e.g., a study area mask?

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
````diff
@@ -0,0 +1,400 @@
+)
+```
+
+# Output Files Structure Naming Conventions
````
Maybe mention that this is strives to be similar to CMIP(5/6) naming convention

CaitlinA replied:
fixed - line 320

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+* Project – Project Name (i.e. SEUG or RangeResil, etc.)
+* SOILWAT2 – Keep as is. Denotes that SOILWAT2 was used to generate these results
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (i.e. IPSL5, CCSM2-1)
+* RCP – Name of which GCM used to drive the GCM (RCP4.5, RCP8.5)
```
Name of experiment under which the GCM was run

CaitlinA replied:
Where does this go?

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+* SOILWAT2 – Keep as is. Denotes that SOILWAT2 was used to generate these results
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (i.e. IPSL5, CCSM2-1)
+* RCP – Name of which GCM used to drive the GCM (RCP4.5, RCP8.5)
+* variantLabel –Variant of the GCM used (i.e. r1i1p1f1)
```
i.e., -> e.g.,

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+* timestep – The time-step of the data contained (i.e. the units of the time dimension (i.e. year, month)).
+* Project – Project Name (i.e. SEUG or RangeResil, etc.)
+* SOILWAT2 – Keep as is. Denotes that SOILWAT2 was used to generate these results
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (i.e. IPSL5, CCSM2-1)
```
i.e., -> e.g.,

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (i.e. IPSL5, CCSM2-1)
+* RCP – Name of which GCM used to drive the GCM (RCP4.5, RCP8.5)
+* variantLabel –Variant of the GCM used (i.e. r1i1p1f1)
+* gn
```
explain that gn represent "native grid" -- which should only be used if the SOILWAT2 output resolution has not been modified

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+will contain variable output for one time period and/or climate model.
+
+The name for each file should be in the format as follows:
+**var_LorA_timestep_Project_SOILWAT2_GCM_RCP_variantLabel_gn_startT_endT.nc**
```
CMIP uses startT-endT instead

dschlaep replied:
Should we also add a `-clim` suffix if the file contains climatologies instead of time-series?

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+   # dummy data for 10 sites, 10 data points (could be 10 vars, 10 times, or 10 depths)
+   someData <- rnorm(100, 7, 30)
+   data1d <- array(someData, c(10, 10))
```
I wasn't interpreting much into this variable name, but then `data3d` suggests that it contains 4-dimensional data collapsed into 3 dimensions (2d space, vertical, time). So, why is the 2-dimensional array called `data1d` and not `data2d` (2d space, vertical)?

CaitlinA replied:
changed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+      crs = crs,
+      force_v4 = TRUE
+    )
+
```
I would like to see a demonstration for each example how to read and recreate the inputs successfully, e.g., something like
```
all.equal(read_data(...), data1d)

tmp <- read_coordinates(...)
all.equal(tmp[["coords_spatial"]], locationsGrid)
all.equal(tmp[["coords_vertical"]], vertical_attributes[["vals"]])
all.equal(tmp[["coords_time"]], time_attributes[["vals"]])
```

CaitlinA replied:
added

CaitlinA replied:
Well I made sure all well values were equal. Not other attributes.

dschlaep replied:
Nice!

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -118,68 +300,125 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
     missval = NULL, prec = "integer")
 
-  # define dimension bounds
+  # define dimension bands ---------------------------------------------------------------
```
these are not "dimension bands" but "dimension bounds", here and throughout

CaitlinA replied:
fixed

dschlaep replied:
line 528 is not yet fixed

dschlaep replied:
R/functions_netCDF.R, line 528: not fixed, change "bands" to "bounds"

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -188,36 +427,79 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
+  }
+
+  if (!is.na(prj)) {
+    ncdf4::ncatt_put(nc, varid = "crs", attname = "proj4",
```
I think we should change how we handle CRS to what CF-1.8 has: http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#grid-mappings-and-projections, e.g., see "Example 5.9. Latitude and longitude on the WGS 1984 datum", and provide our CRS string as attribute name `crs_wkt` instead (see http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#use-of-the-crs-well-known-text-format)

CaitlinA replied:
added

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -230,268 +512,307 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
-  time_bounds, var_attributes, global_attributes, file, force_v4 = TRUE,
-  overwrite = FALSE) {
+
+populate_netcdf_from_array <- function(file, data, var_names = NULL,
```
Could we add a verbose = FALSE argument and wrap the messages into `if (verbose) message(...)`?

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,98 +1,278 @@
-  var_attributes, global_attributes, file, force_v4 = TRUE,
-  overwrite = FALSE) {
 
+create_empty_netCDF_file <- function(data, has_T_timeAxis = FALSE,
```
Could we add a verbose = FALSE argument and wrap the messages into `if (verbose) message(...)`?

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -230,268 +512,307 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
+#'    parent_mip_era = "CMIP5",
+#'    parent_experiment_id = "RCP45",
+#'    parent_source = "CanESM2",
+#'    parent_variant_label = "r1i1p1f1",
```
CMIP5 used rip as variant label, i.e., it didn't include information about the forcing (index) in the variant label -- CMIP6 uses ripf as it added the forcing index

CaitlinA replied:
added

dschlaep replied:
lines 801 and 804 continue to be inconsistent

dschlaep replied:
R/functions_netCDF.R, line 802 (`parent_variant_label = "r1i1p1f1"`) is still in conflict with line 799 (parent_mip_era = "CMIP5").

CaitlinA replied:
fixed

dschlaep commented on .Rbuildignore:
```diff
@@ -20,3 +20,4 @@
 # CRAN stuff
 ^cran-comments\.md$
 ^NEWS\.md$
+^vignettes/dummynetcdf\.nc$
```
Need to ignore `dummynetcdf.nc` wherever it was created and not cleaned up including rSW2analysis/ or rSW2analysis/R/ depending from which location `devtools::run_examples()` was run -- or at least until the example code cleans up

CaitlinA replied:
added to ignore file

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,98 +1,278 @@
+#'      force_v4 = TRUE,
+#'      overwrite = TRUE
+#'    )
+#'
```
add `unlink(outFileName)` to clean up

CaitlinA replied:
I have a question about this

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -230,268 +512,307 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
+#'      crs = crs,
+#'      force_v4 = TRUE
+#'    )
+#'
```
add `unlink(outFileName)` to clean up

CaitlinA replied:
I have a question about this

CaitlinA replied:
fixed

CaitlinA committed:
Major improvement to non-gridded, orthogonal time series functionality

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,98 +1,278 @@
+#'    this argument can be used to define the calculation period of the variable(s).
+#' @param vert_bounds A numeric string that continuously lists the lower and upper
+#'    bounds of each vertical dimension value.
+#' @param var_attributes A list of named character strings defining the variables
```
I think we should encourage using CMIP6 standard names where available and link to variable lists, e.g., the  [day](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_day.json) and [Eday](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_Eday.json) tables for time-varying variables and to [fx](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_fx.json) and [Efx](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_Efx.json) for time-invariant variables.

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -230,268 +512,307 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,
+#'
+#' @param file A character string. File path to a pre-created, empty netCDF.
+#' @param data A numeric array.
+#' @param var_name A character string equal to the length of the number of variables.
```
same as above, link to CMIP standard variable names

CaitlinA replied:
fixed

CaitlinA replied:
Not sure what this refers to.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+
+### Variable Attributes
+
+A named list describing the variable data. The *name* and *units* fields are required.
```
As for the function documentation, I think we should encourage using CMIP6 standard names where available and link to variable lists, e.g., the  [day](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_day.json) and [Eday](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_Eday.json) tables for time-varying variables and to [fx](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_fx.json) and [Efx](https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_Efx.json) for time-invariant variables.

CaitlinA replied:
fixed

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,400 @@
+This vignette is an overview of how to use the functions in the rSW2analysis package to
+convert arrays of rSOILWAT2 and SOILWAT2 output data into netCDFs that adhere to
+[CF conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#_orthogonal_multidimensional_array_representation).
+
```
maybe also add that much is following CMIP5/6 as well

CaitlinA replied:
fixed

CaitlinA committed:
Remove has_time_Central arguments from code

Reconfigure how function determines if array is organized by variable, 
time, depth (now looks at length of vals in attributes)

CaitlinA committed:
Add crs_attributes list as argument, including crs_wkt

CaitlinA committed:
Major fixed to populate_netcdf_From_array

Also:
*updates to documention
*update to vignette
*better error handling

CaitlinA committed:
Finish read_netcdf_to_array function

* Major work on read_netcdf_to_array
* Fixes to the vignette
* Fixxes to populate_netcdf_From_array

CaitlinA committed:
Fixes based on review

* Mostly changes to vignette
* Some changes to function documention
* Added verbose = FALSE argument to functions

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +401,238 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+                         start = c(1, 1), count = c(2L, var_chunksizes[2])))
+  } else {
+    try(ncdf4::ncvar_put(nc, varid = "lon",
+                         vals = locations$X_WGS84,
```
Earlier, we converted `loc <- sp::SpatialPoints(locations)`. Thus, we could use `sp::coordinates(locations)` instead of assuming a specific and undocumented naming of the xy-columns of `locations` - that also only make sense if the crs is WGS84

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +654,374 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
-      attval = as.character(prj))
+  # The end --------------------------------------------------------------
+
+  if(verbose == TRUE) {
```
`if (verbose)` is sufficient because `verbose` is logical

CaitlinA replied:
fixed

CaitlinA committed:
Fixes and improvements  to netcdf code

* Added 'sf' workflow to functions
* Added checks in vignette
* Other small fixes

CaitlinA committed:
Update DESCRIPTION

CaitlinA committed:
Delete dummynetcdf.nc

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+Additionally, we strongly recommend defining the CRS in the crs_wkt format, which
+is preferred by [netCDF convention] (http://cfconventions.org/cf-conventions/cf-conventions.html#use-of-the-crs-well-known-text-format).
+
+The *wkt* function in the raster package can be used to generate wkts.
```
Please mention equivalents:
- `sf` package: `cat(sf::st_crs(4326)$wkt, "\n")`
- `sp` package: `cat(sp::wkt(sp::CRS(SRS_string = "EPSG:4326")), "\n")` -- if `rgdal` is build against PROJ >= 6 and GDAL >= 3

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+)
+
+locationsSP <- sp::SpatialPoints(locations)
+sp::proj4string(locationsSP) <- sp::CRS(crs_attributes$proj)
```
why do this in two lines 260-261 instead of one:
`locationsSP <- sp::SpatialPoints(coords = locations, proj4string = sp::CRS(crs_attributes$proj))

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+sp::proj4string(locationsSP) <- sp::CRS(crs_attributes$proj)
+WKTex <- raster::wkt(locationsSP)
+
+crs_attributes <- c(crs_attributes, crs_wkt = WKTex)
```
a more efficient approach instead of copying `crs_attributes`:
`crs_attributes[["crs_wkt"]] <- WKTex`

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+    parent_mip_era = "CMIP5",
+    parent_experiment_id = "RCP45",
+    parent_source = "CanESM2",
+    parent_variant_label = "r1i1p1f1",
```
Same comment as for the example code in the function documentation:
it is impossible to have a `ripf`-variant for the `CMIP5`-era; variants in the `CMIP5`-era are coded in the format `rip`.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+    projection = 'Geographic',
+    grid = 'WGS84',
+    grid_label = "gn",
+    nominal_resolution = "10 km", # \code{\link{populate_netcdf_from_array}}
```
I don't think a user understands what `\code{\link{populate_netcdf_from_array}}` means. Comment instead: "# calculated with function populate_netcdf_from_array()" - or similar

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+calls for the field *nominal resolution* to be contained in the global attributes.
+The function \code{calculate_nominal_resolution} will calculate the nominal resolution
+of your gridded data for you. However, you need to provide a raster of the grid,
+the lat/long locations, and the area that each cell in the grid represents.
```
Here and throughout, I think we need to be careful about whether we refer to data in lat/long coordinate pairs or long/lat coordinates.

See [R spatial follows GDAL and PROJ development]{https://www.r-spatial.org/r/2020/03/17/wkt.html}:
- PROJ respects authorities (such as EPSG) for determining whether coordinate pairs refer to longitude-latitude (such as 3857), or latitude-longitude (such as 4326)
- GDAL offers the ability to handle coordinate pairs authority-compliant (lat-long for 4326), or “traditional” GIS-compliant (long-lat for 4326)
- `sf::st_axis_order`
- Workflows using sp/rgdal should expect “GIS style” axis order to be preserved

CaitlinA replied:
Hmm - Could you point to where you want this changed? Because this file is now outdated, I am no longer able to find where you refer to.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+where:
+
+* var – Shorthand name for variable that is contained in files (i.e. ppt for precipitation, WDD for wet degree days)
+* LmonOrAmon – Land (L) or atmospheric (A) data
```
Many of our variables are in the CMIP6 table "E" (extended) -- I don't know whether we want to override those table assignments and make our own based on the <realm_id> or stick with the CMIP6 standards, or drop the "realm" part altogether?

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+
+* var – Shorthand name for variable that is contained in files (i.e. ppt for precipitation, WDD for wet degree days)
+* LmonOrAmon – Land (L) or atmospheric (A) data
+* timestep – The time-step of the data contained (i.e. the units of the time dimension (e.g. year, month)).
```
The previous "table" part already contains the time step information if formatted as in line 343, e.g., the "mon" means "monthly" -- see https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/CMIP6_table_id.json

The <table_id> is for the most part a concatenation of the <realm_id> and the <frequency_id>, see https://github.com/WCRP-CMIP/CMIP6_CVs/blob/master/

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+* SOILWAT2 – Keep as is. Denotes that SOILWAT2 was used to generate these results
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (e.g. IPSL5, CCSM2-1)
+* RCP – Name of experiment under which the GCM was run (e.g. RCP4.5, RCP8.5)
+* variantLabel – Variant of the GCM used (e.g. r1i1p1f1)
```
Comment as earlier, CMIP5 used rip, CMIP6 uses ripf. The RCPs listed in line 349 are CMIP5 experiments suggesting that this should be a rip (and not ripf) <variant_id>

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,333 +1,420 @@
+#'    parent_mip_era = "CMIP5",
+#'    parent_experiment_id = "RCP45",
+#'    parent_source = "CanESM2",
+#'    parent_variant_label = "r1i1p1f1",
```
as commented elsewhere: the combination of ripf and CMIP5 does not exist

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,333 +1,420 @@
+      if(isGridded) grid_halfres <- loc@grid@cellsize / 2
+    }
+    if(inherits(loc, 'sf')) {
+      xvals <- sort(unique(st_coordinates(locations)[,1]))
```
lines 298-299: --> sf::st_coordinates

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,333 +1,420 @@
-  if (!is.na(prj)) {
-    ncdf4::ncatt_put(nc, varid = "crs", attname = "proj4",
-      attval = as.character(prj))
+  # Note: xvals should be organized from west to east, yvals from north to south
```
Why are the `yvals [sorted] from north to south`, i.e., decreasingly? This seems counter-intuitive to me. I don't see anything in the CF standard that this is required: http://cfconventions.org/cf-conventions/cf-conventions.html#latitude-coordinate

CaitlinA replied:
You are correct. This is just legacy code.

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,333 +1,420 @@
-    # avoid "_FillValue" error in older versions of `raster` package
-    stopifnot(utils::packageVersion("raster") >= "2.9.1")
+  # Time dimension setup & info ----------------------------------------------------------------
+  if(has_T_timeAxis & !is.null(time_bounds)) {
```
`if` expects a single value (see `_R_CHECK_LENGTH_1_CONDITION_`) --> change `&` to `&&`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +424,248 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+  if(!isGridded){
+   latvar <-  ncdf4::ncvar_def(name = 'lat', units = 'degrees_north',
+                               dim = list(idim), chunksizes = var_chunksizes[1], missval = NAflag,
+                               longname = 'site latitiude', prec = ncdf4_datatype)
```
line 515: longname is misspelled
line 515 and 519: why do we add "site" to the lat/long longnames?

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +424,248 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
-  nc_dimvars <- list(lonbnddef, latbnddef)
-  if (has_time_central) {
+  if (has_T_timeAxis) {
+      tbnddef <- ncdf4::ncvar_def(name = "time_bnds", units = "",
```
Lines 540 and 546: Since the names of the "time" and "vertical" dimensions are user provided, shouldn't the associated bounds reflect their names accordingly? Or, we fix the "time" and "vertical" dimension names as we did for longitude/latitude.
As it is now, a user could define a "vertical" axis named "height" and the function would create associated bounds as "depth_bnds" - confusing to future users of that netCDF.

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+populate_netcdf_from_array <- function(file, data, var_names = NULL,
+                                       has_T_timeAxis, has_Z_verticalAxis,
+                                       isGridded = TRUE, grid = NULL, locations,
+                                       force_v4 = TRUE, verbose = FALSE) {
```
Function `populate_netcdf_from_array` does not use the argument `force_v4` --> remove

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +424,248 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+  }
+
+  # add coordinate system attributes --------------------------------------------
+  if(!is.null(grid)){
```
lines 651-655: what is this for? I don't see that `prj` is used anywhere

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +424,248 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+
   ncdf4::ncatt_put(nc, varid = 0, attname = "created_by",
     attval = paste0(R.version[["version.string"]], ", R packages ",
       if (requireNamespace("rSFSW2")) {
```
lines 671-673: this is now package `rSW2analysis`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
-      ncdf4::ncatt_put(nc, varid = var_names[k], attname = natt,
-        attval = var_attributes[[natt]][k])
+  stopifnot(requireNamespace("ncdf4"))
+  if (force_v4) {
```
lines  864-867: not needed in this function --> remove

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+  if(isGridded) {
+    if(inherits(locations, 'sf')) {
+      loc <- sf::st_geometry(locations)
+      st_crs(loc) <- crs 
```
--> `sf::st_crs`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+      st_crs(loc) <- crs 
+      loc <- as(loc, Class = 'Spatial')
+    } else { # either assumes its spatial class or data.frame
+      loc <- sp::SpatialPoints(locations)
```
lines 907-908: simplify to `loc <- sp::SpatialPoints(locations, proj4string = sp::CRS(crs))`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
     }
+    
+    if(is.null(grid)){       # make grid
+      sp::gridded(loc) = TRUE
```
use `<-`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+    # ---------------------------------------------------------------------
+    # NOT GRIDDED ---------------------------------------------------------
+    # ---------------------------------------------------------------------
+    if(isGridded == FALSE) {
```
--> `if (!isGridded) {`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+    }
+    
+    #  GRIDDED --------------------------------------------------------------
+    if(isGridded == TRUE) {
```
--> `if (isGridded) {`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+      
+      # more checks based on dim of var ----------------------------------------
+      
+      if(has_T_timeAxis & !has_Z_verticalAxis) stopifnot(nc_var_dims[3] == nn)
```
lines 1004-1005: replace `&` with `&&`
--> even better, combine lines 1004 and 1005 `if (isTRUE(xor(has_T_timeAxis, has_Z_verticalAxis)))`

CaitlinA replied:
Never seen xor. Cool!

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+    
     # Flush this step to the file so we dont lose it
     # if there is a crash or other problem
     ncdf4::nc_sync(nc)
```
line 1063: nc_sync is no longer inside a loop and now defeats purpose --> move it inside the loops or remove

CaitlinA replied:
removed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+  # figure out dimensionality
+  nc_dims <-  attributes(nc$dim)$names #dims of netcdf
+  
+  if('site' %in% nc_dims) isGridded = FALSE
```
lines 1114-1115: use `<-` instead of `=` for assignments

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+  if('site' %in% nc_dims) isGridded = FALSE
+  if('lat' %in% nc_dims) isGridded = TRUE
+  
+  has_T_timeAxis <-  ifelse('time' %in% nc_dims, TRUE, FALSE)
```
lines 1117 and 1118 (equivalently): --> `has_T_timeAxis <- "time" %in% nc_dims`

dschlaep replied:
also, see comment to line 480: the names of the time and vertical dimensions are user specified; however, here we assume a specific name. This doesn't work when a user names the vertical dimension for instance "height".
--> I think the best option is to fix the names of the these two dimensions as we do for longitude and latitude

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -337,105 +424,248 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+
+  # time dimension
+  if(has_T_timeAxis) {
+      tdim <-  ncdf4::ncdim_def(name = time_names,
```
I think we should fix the names of the time and vertical dimensions. See comments at several places later in the file

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+  nc_var_names <- attributes(nc$var)$names
+  nc_var_names <- grep('bnds|crs|lat|lon', nc_var_names, invert = TRUE, value = TRUE)
+  
+  if(isGridded == TRUE) {
```
--> `if (isGridded) {`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+  if(isGridded == TRUE) {
+    
+    # location stuff  ---------------------------------------------------------
+    loc <- sp::SpatialPoints(locations)
```
what if locations is a `sf` object?

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+    # location stuff  ---------------------------------------------------------
+    loc <- sp::SpatialPoints(locations)
+    # Create grid from location values ----------------------------------------
+    sp::gridded(loc) = TRUE
```
<-

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+    nc_var <- ncdf4::ncvar_get(nc, nc_var_names[1])
+    nc_var_dims <- dim(nc_var)
+    
+    locDims <- c(nc_var_dims[1] * nc_var_dims[2])
```
`c()` is unnecessary --> remove

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+    
+    nc_var <- ncdf4::ncvar_get(nc, nc_var_names[1])
+    nc_var_dims <- dim(nc_var)
+    newArray <- array( , dim = c(locDims, 
```
dangling comma

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+    newArray <- array( , dim = c(locDims, 
+                                 length(nc_var_names)))
+    
+    for(k in 1:length(nc_var_names)) { 
```
replace `1:length(nc_var_names)` with `seq_along(nc_var_names)`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+        
+        for(i in seq(dimtz)) {
+          
+          if(isGridded == TRUE) v <- c(nc_var[, , i, z]) else v <- c(nc_var[, i, z])
```
-> `if (isGridded)`

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+      
+      for(i in seq(dimtz)) {
+        
+        if(isGridded == TRUE) v <- c(nc_var[, , i]) else v <- c(nc_var[, i])
```
-> if (isGridded)

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -533,11 +1096,127 @@ read_netCDF_to_raster <- function(x, ...) {
+    }
+  }
 
+  return(newArray)
```
`return` is not needed --> `newArray`

dschlaep reviewed (changes requested):
I added some comments to address and a few questions (e.g., file naming) to discuss.
This branch needs to be updated from the main branch and code lints and spelling issues should be fixed (so that `devtools::check(env_vars = c(NOT_CRAN = "true"))` passes and doesn't add notes or warnings relative to what the main branch has.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+depths_max <- c(5, 10, 15) # maximum depth of each layer of depth
+
+## depth bounds
+depth_bounds <- c(0, 5, 5, 10, 10, 15) # min and max of each nlayer of depth depth
```
comment repeats `depth`?

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,599 @@
+
+The user must define the bounds and values for the depth dimension.
+
+For depth values, we use the maximum depth value.
```
line 146: Why are we using the lower soil layer depths as values for the vertical dimension? Depending on the SOILWAT2 variable, the SOILWAT2 output may refer to the entire layer width, e.g., VWC (in which case the values should reflect the layer midpoints), or to a specific depth, e.g., soil temperature (in which case we shouldn't have bounds...).

CaitlinA committed:
Fix simple comments from pull request

CaitlinA committed:
hard coded time and dimension name info

CaitlinA committed:
Updates and fixes based on review

CaitlinA committed:
Merge branch 'master' into ImproveNetCDFFunctionality

CaitlinA committed:
fix merge

CaitlinA committed:
Lintr and spelling fixes

CaitlinA committed:
small lint fixes

CaitlinA committed:
Updating functions_Raster.

I messed up merge from master apparently.

CaitlinA committed:
small lints

CaitlinA commented:
Hi,

I fixed all the lintr and spelling checks in the new files.

Somehow, I messed up the merge from master and had to copy/paste functions_raster.R.

I now see no new errors/warnings/notes when running `devtools::check(env_vars = c(NOT_CRAN = "true"))`
on this branch vs. master.

dschlaep commented:
@CaitlinA Great, thanks! I will look at it in the next few days.

dschlaep commented on DESCRIPTION:
```diff
@@ -45,6 +45,10 @@ Suggests:
-    colorspace
+    colorspace,
+    knitr,
+    rmarkdwon,
```
Package "rmarkdown" is misspelled -- this also shows up in package checks

CaitlinA replied:
fixed!

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -1,106 +1,417 @@
-
-#' Write raster where a third dimension represents time to a netCDF file
+#'  Create a structured, empty netcdf file
```
I think it would be easier to use consistent terminology. Maybe use "netCDF" throughout instead of "netcdf".

CaitlinA replied:
fixed

dschlaep commented on R/functions_netCDF.R:
```diff
@@ -110,113 +421,254 @@ create_netCDF_from_raster_with_variables <- function(x, time_bounds,  #nolint
     attval = paste0(R.version[["version.string"]], ", R packages ",
-      if (requireNamespace("rSFSW2")) {
-        paste0("rSFSW2 v", utils::packageVersion("rSFSW2"), " and ")
+      if (requireNamespace("rSW2analysis")) {
```
Since this is package `rSW2analysis`, the `if (requireNamespace("rSW2analysis")) ...` is no longer needed

CaitlinA replied:
deleted

dschlaep commented on dev_checks.R:
```diff
@@ -0,0 +1,31 @@
+library(roxygen2)
```
I don't see a reason why the file "dev_checks.R" should be on the repository

dschlaep commented on inst/WORDLIST:
```diff
@@ -1,75 +1,103 @@
-###---- Spell checking
```
I personally preferred the previously organized wordlist, but apparently `spelling::update_wordlist` doesn't think so.

CaitlinA replied:
Should we change it back?

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,603 @@
+* RCP – Name of experiment under which the GCM was run (e.g. RCP4.5, RCP8.5)
+* variantLabel – Variant of the GCM used (e.g. r1i1p1 (CMIP5 format), r1i1p1f1 (CMIP6 format))
+* gn - grid_label. "gn" is the "native grid", which should only be used if the SOILWAT2 output resolution has not been modified.
+* startT – lower time bound or start of the time period contained. Units should be the date (YearMonthDay, XXXXYYZZ).
```
A more usual representation of a date in this format would be YYYYMMDD (see, e.g., https://en.wikipedia.org/wiki/ISO_8601)

CaitlinA replied:
makes sense. fixed.

dschlaep commented on vignettes/netCDF_Functionality.Rmd:
```diff
@@ -0,0 +1,603 @@
+
+If the data is on a grid, either a gridded raster file or a table of coordinates
+can be provided. If the former, the coordinates will be extracted from the grid.
+If the later, the coordinates need to be ordered from least to greatest. For example,
```
I'm not sure what "coordinates ... ordered from least to greatest" means. Since these are numbers maybe from smallest to largest, but probably more usual and useful would be to remind that the orientation is east (for longitude) and north (for latitude), e.g., https://epsg.org/crs_4326/WGS-84.html and https://en.wikipedia.org/wiki/Longitude, and should be ordered accordingly?

CaitlinA replied:
Tried to clarify

dschlaep reviewed (changes requested):
This is coming nicely together.

I went through all comments and marked them as resolved if so. I added a few more comments and there are some that I believe to be worth to still be addressed, including
- https://github.com/DrylandEcology/rSW2analysis/pull/1#discussion_r494363633
- https://github.com/DrylandEcology/rSW2analysis/pull/1#discussion_r498925593
- https://github.com/DrylandEcology/rSW2analysis/pull/1#discussion_r498926443
- https://github.com/DrylandEcology/rSW2analysis/pull/1#discussion_r485113614
- https://github.com/DrylandEcology/rSW2analysis/pull/1#discussion_r494373370

CaitlinA committed:
Resolving pull review comments

* Added clarity to documentation
* Fixed small wording and syntax errors.

CaitlinA commented:
@dschlaep I looked over your requested changes and had a few questions. Thanks!

dschlaep reviewed (changes requested):
@CaitlinA Thanks for the updates!

Two new points
1) Updates of GDAL/PROJ to v3+/6+ changed how `crs` work (e.g., https://rgdal.r-forge.r-project.org/articles/CRS_projections_transformations.html): 
> PROJ strings to encode CRS should be considered as a a legacy solution. Using a AUTHORITY:CODE or WKT representation is the recommended way” (https://gdal.org/api/ogrspatialref.html).
> For these reasons, sf and sp are changing from PROJ.4 strings representing coordinate reference systems to WKT2:2019 strings

I think we should follow suit, drop the `proj` element, and require instead `crs_wkt`, e.g., `sf::st_crs("OGC:CRS84")$Wkt`.

2) With a `grid mapping` variable and `crs_wkt` element, our variables require `grid_mapping` (in the extended version) as additional attribute according to CF standards (see, e.g., Example 5.11. Latitude and longitude on the WGS 1984 datum + CRS WKT; http://cfconventions.org/cf-conventions/cf-conventions.html#grid-mappings-and-projections): e.g., `grid_mapping = "crs: latitude, longitude"`

CaitlinA committed:
Fixes to latest request

* removed references to proj
* switched crs class defintion to crs_wkt
* add variable attribute grid map to netcdf if gridded

CaitlinA commented:
@dschlaep 

Made fixes based upon your request. 

Questions:

* Do we need to update package version requirements in the NAMESPACE?
* the `proj4string` argument is still used (i.e. line 208) to define the CRS of locations object of class Spatial. As far as I could tell this argument is still used, even when the CRS is defined as a crs_wkt. What do you think?
* I automatically assign the `grid_mapping` variable attribute in the `create_netCDF_from_array` function, if the data is gridded. Do you think this is correct or should it be a user set attribute in the `var_attributes` list?

dschlaep commented:
Thanks for the updates.

- it would be ideal to update package version requirements in the `NAMESPACE`, e.g., `sp` requires v1.4.2 if we use its `wkt` (https://stat.ethz.ch/CRAN/web/packages/sp/news.html)
- for consistency, we should use `rSW2st::as_points()` to convert any `locations` inputs to a spatial object of class `sf`; and use `sf::st_crs()` for any handling of crs objects (e.g., lines 206 `crs <- locations@proj4string` and 208 `crs <- sf::st_crs(locations)[[1]]` collapse and simplify to `crs <- sf::st_crs(locations)$Wkt`; alternatively, we simply require that `crs_attributes[["crs_wkt"]]` is a valid crs. On that point: I don't understand why we have lines 272-281 that re-assign the crs to a spatial `locations` object from which we extracted it in lines 206/208 -- unless crs was defined via `crs_attributes[["crs_wkt"]]` in which case it could be different and we would need to transform `locations` instead of overwriting the crs.)
- it seems to me that the best solution for the `grid_mapping` attribute is to make it a required field in the `var_attributes` if there is a `crs_attributes` -- because the value of `grid_mapping` is dependent on the `crs` and I don't think it is worth attempting to calculate the value of `grid_mapping` based on `crs_attributes`; I anticipate that we will likely have a range of crs types including WGS84, LCC, and AEA.
- re-reading the CF-convention document makes me think that the copy-pasted `grid_mapping` example 5.11  `grid_mapping = "crs: latitude, longitude"` is incorrect; it is defined as "blank-separated list of words" (http://cfconventions.org/cf-conventions/cf-conventions.html#grid-mappings-and-projections) as used in their examples 5.10, 5.12, and 5.13. Their example should be `grid_mapping = "crs: latitude longitude"`; however, our coordinate variable names are `lat` and `lon`, thus, it seems that it should be for our use: `grid_mapping = "crs: lat lon"`

CaitlinA commented:
@dschlaep 

> for consistency, we should use rSW2st::as_points() to convert any locations inputs to a spatial object of class sf; and use sf::st_crs() for any handling of crs objects (e.g., lines 206 crs <- locations@proj4string and 208 crs <- sf::st_crs(locations)[[1]] collapse and simplify to crs <- sf::st_crs(locations)$Wkt; alternatively, we simply require that crs_attributes[["crs_wkt"]] is a valid crs. 

Line 205 - 209 are set up to attempt to extract the CRS from the locations object if the crs_attributes[["crs_wkt"]] is _not_ available. It only works if the locations object is of class 'Spatial' or 'sf' and the projection is defined. The locations objects can also just be a data.frame, so no spatial data attached nor is it required by the function.

Honestly, I could maybe should just eliminate this check altogether because defining crs_attributes[["crs_wkt"]] is a _requirement_ of the function and it will fail down the road if this is not filled out.

Lines 271 - 280 are the critical element of the functions for assuring the spatial information is correctly defined. Here I advocate for keeping sf and sp functions separate, as the sf package is not yet as ubiquitous as the sp package. I think it is good to have both options.

dschlaep commented:
> Lines 271 - 280 are the critical element of the functions for assuring the spatial information is correctly defined. Here I advocate for keeping sf and sp functions separate, as the sf package is not yet as ubiquitous as the sp package. I think it is good to have both options.

The function `rSW2st::as_points()` provides an abstraction so that other functions do not have to repeat if (sp) ... else if (sf) ... all the time. They can pass their input to this function which takes care of the necessary stuff to handle suitable non-spatial, sp, and sf objects.

Another problem is that the current code fails to achieve the goal of "assuring the spatial information is correctly defined". If a spatial `locations` argument is provided that does not agree with the `crs_wkt`, then lines 271 - 280 overwrite instead of transform the spatial object. As example:
```
#--- Inputs
x <- data.frame(
  X_WGS84 = c(rep(-124.5938, 5), rep(-124.5312, 5)),
  Y_WGS84 = rep(c(47.90625, 47.96875, 48.03125, 48.09375, 48.15625), 2)
)

# Create a spatial points in (web) pseudo-mercator projection
locations <- sf::st_transform(
  rSW2st::as_points(x, to_class = "sf", crs = 4326), 
  crs = 3857
)

# Define crs as WGS84 -- in disagreement with the crs of `locations`
crs_attributes <- list(crs_wkt = sf::st_crs(4326)$Wkt)

#--- Then the function currently does
# Line 204
crs <- crs_attributes[["crs_wkt"]]

# Lines 274-275: Note the warning issued by `sf::st_crs()`!
loc <- sf::st_geometry(locations)
sf::st_crs(loc) <- crs

# Lines 304-305 & 319: Produce nonsensical coordinate values given the declared crs
sort(unique(sf::st_coordinates(locations)[, 1]))
sort(unique(sf::st_coordinates(locations)[, 2]))

crs_attributes[["crs_wkt"]]
```

CaitlinA committed:
Updates to netCDF functions

* checks for CRS matching
* CRS defined with rSW2st::as_points function
* sf class used for locations data in more instances
* grid_mapping is a required var_attribute when gridded

CaitlinA commented:
@dschlaep 

I made some changes. Please check it out.

* added a check to see if the stated CRS and the CRS of spatial objects match
* define the CRS and the class of locations with the as_points functions
* added `grid_mapping` into the `var_attributes` list

dschlaep commented:
Great, thanks!

> added a check to see if the stated CRS and the CRS of spatial objects match

`crsL != crs` (e.g., line 218 and others) doesn't necessary do what I guess you wanted it to do. `crs` is the user provided input from `crs_attributes[["crs_wkt"]]` (line 202) - hopefully a character string representing a valid WKT; ideally, we would test that this is so. 
- If `crsL` is an object of class `crs` from the `sf` package (in case of line 214), then the test for unequal CRSs will always return TRUE whether or not `crsL` and `crs` represent the same CRS.
- If `crsL` is a character string representing a WKT produced by the `sp` package (in case of line 210), then the two character strings `crsL` and `crs` may still  differ even if they are both representing the same valid WKT due to white space, line returns, etc.

A better way to test for whether two CRSs are representing the same is via the `crs`-class comparisons, i.e., `sf::st_crs(crsL) != sf::st_crs(crs)`

CaitlinA committed:
Small fix to CRS check

CaitlinA commented:
Duh. Fixed that error.

I didn't add a check to see if it a valid WKT. I did fid this function, but not sure if we want another dependency or if it will do exactly what we want: https://rdrr.io/cran/wicket/man/validate_wkt.html

dschlaep commented:
I agree that we don't need a dependency on `wicket` -- also `wicket::validate_wkt()` only considers the WKT representation of spatial objects and not the WKT representation of CRSs:
`wicket::validate_wkt(sf::st_crs(4326)$Wkt)`
>  is_valid                                               comments
1    FALSE Object could not be recognised as a supported WKT type

What about
```
# Check that `crs_wkt` represents a valid CRS
tmp <- try(sf::st_crs(crs_attributes[["crs_wkt"]]), silent = TRUE)
if (!inherits(tmp, "crs") || tmp == sf::NA_crs_) {
  stop("`crs_attributes[[\"crs_wkt\"]]` does not represent a valid CRS.")
}
```

example cases:
```
# Invalid because `st_crs()` produces `NA_crs_`
crs_attributes <- list(crs_wkt = 5)

# Invalid because `st_crs()` produces error
crs_attributes <- list(crs_wkt = "GEOGCS['WGS 84',DATUM['WGS_1984'")

# Valid
crs_attributes <- list(crs_wkt = sf::st_crs(4326)$Wkt)
```

CaitlinA committed:
Update functions_netCDF.R

* check for valid CRS

CaitlinA commented:
@dschlaep Good idea. Updated

dschlaep commented:
Great! Let's move to close this up.

I ran `devtools::check()` and it failed due to `grid_mapping` not being used and due to code style issues. 
The branch should also be updated from the master branch to enable to be closing the PR.

CaitlinA committed:
Merge branch 'master' into ImproveNetCDFFunctionality

CaitlinA committed:
Small fixes for checks

CaitlinA commented:
Ok I merged the master branch into the dev branch. The only conflicts were in the DESCRIPTION file, and I think fixed all conflicts correctly. Please double check.

Fixed the lintr and grid mapping issues as well. I was running devtools::check(env_vars = c(NOT_CRAN = "false")) before.

Should be ready to close pull request and merge.
________________________________
From: Daniel R Schlaepfer <notifications@github.com>
Sent: Tuesday, November 10, 2020 3:04 PM
To: DrylandEcology/rSW2analysis <rSW2analysis@noreply.github.com>
Cc: Andrews, Caitlin M <candrews@usgs.gov>; Mention <mention@noreply.github.com>
Subject: [EXTERNAL] Re: [DrylandEcology/rSW2analysis] Improve net cdf functionality (#1)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



Great! Let's move to close this up.

I ran devtools::check() and it failed due to grid_mapping not being used and due to code style issues.
The branch should also be updated from the master branch to enable to be closing the PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDrylandEcology%2FrSW2analysis%2Fpull%2F1%23issuecomment-724994436&data=04%7C01%7Ccandrews%40usgs.gov%7Ccde167e5386a4054d9fb08d885c4ce2e%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637406427608020303%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=QtQ8qWA5BI1Irr03%2FN4Z%2F7ErWlD%2B2uOfSL6C8Sr9MMw%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?%20url=http%20s%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FADGR57Z4TRAHNVRFC4ORZFDSPG2G3ANCNFSM4QNV2MKA&data=04%7C01%7Ccandrews%40usgs.gov%7C15383fa5363749bfb5cd08d885c498ed%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637406427149012590%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=djr4I6uRlhBfrybf6rNb0SJB0hkniw00ZxCSxuUiP8o%3D&reserved=0>.

dschlaep commented:
I've run the vignette and looked at the output, e.g., with `ncdump` and via https://pumatest.nerc.ac.uk/cgi-bin/cf-checker.pl. A couple issues remain:

- The resulting `grid_mapping` variable attribute is incorrectly formatted. It ends up as `grid_mapping = 'grid_mapping = "crs: lat lon"'` instead of, correctly, as `grid_mapping = "crs: lat lon"` (see http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#grid-mappings-and-projections)

- The `unit` variable attribute "number of days" is not a recognized unit by `udunits2` (see http://cfconventions.org/cf-conventions/cf-conventions.html#units); this should be `day` (for instance, compare with variable `agesno` as in https://github.com/PCMDI/cmip6-cmor-tables/blob/master/Tables/CMIP6_Eday.json)

- the many-variable example `2dExample`: 
    * the `long_name` variable attribute is missing
    * the `grid_mapping` variable attribute contains a value only for the first variable
```
	double VarDD_1(lat, lon) ;
		VarDD_1:units = "number of days" ;
		VarDD_1:_FillValue = -1.7e+308 ;
		VarDD_1:grid_mapping = "grid_mapping = \"crs: lat lon\"" ;
	double VarDD_2(lat, lon) ;
		VarDD_2:units = "number of days" ;
		VarDD_2:_FillValue = -1.7e+308 ;
		VarDD_2:grid_mapping = "NA" ;
	double VarDD_3(lat, lon) ;
		VarDD_3:units = "number of days" ;
		VarDD_3:_FillValue = -1.7e+308 ;
		VarDD_3:grid_mapping = "NA" ;
...
```

- the individual-site example `3dNoGridExample`
    * The variable `site` contains `site_id` as units which is not recognized by `udunits2`. This is a dimensionless unit and should be `1`.

- the space&time example `4dGridExample`
    * `WARN: (2.4): space/time dimensions appear in incorrect order` (i.e., space/time dimensions are listed in the order T,Z,Y,X)  -- instead ours are incorrectly: `	double WatYrWDD(vertical, time, lat, lon) ;`

CaitlinA committed:
Fixes to var attibutes

* fixed quotes around grid_mapping example

* fixed insertiions of var_Attributes

* fixed units of idim

CaitlinA committed:
Switch T and Z axis

- switched t and z axis order (if both are present) to conform to netCDF CF1.8 convention.
- other minor changes and fixes

CaitlinA commented:
@dschlaep I made changes addressing these errors, and re-ran all example netCDFs through the python tool. Please check it out!

dschlaep commented:
Nicely resolved!

It seems that the remaining error reported by the cf-checker is due to that our attribute reports "CF-1.8" while they check up to v1.7. That should be ok for us.

I suggest that you close and merge this pull request; and then one of us moves the code to `rSW2st` (so that we can use the netCDF functionality also on the upstream side of `rSFSW2`).
